shared/src/Datadog.Trace.ClrProfiler.Native/crashhandler.cpp

```diff
@@ -0,0 +1,330 @@
+#include "crashhandler.h"
+#include <string>
+#include "util.h"
+#include "WerApi.h"
+
+namespace datadog::shared::nativeloader
+{
+    // Get the path of the current module
+    std::wstring GetCurrentDllPath()
+    {
+        wchar_t path[MAX_PATH];
+        HMODULE hm = NULL;
+
+        if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
+            GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
+            (LPCWSTR)&GetCurrentDllPath, &hm) == 0) {
+            return std::wstring();  // Failed to get module handle
+        }
+
+        if (GetModuleFileNameW(hm, path, MAX_PATH) == 0) {
+            return std::wstring();  // Failed to get module filename
+        }
+
+        return std::wstring(path);
+    }
+
+    // Capture all the environment variables starting with "DD_"
+    void GetDatadogEnvironmentBlock(WCHAR*& environmentVariables, int32_t& length)
+    {
+        auto envStrings = GetEnvironmentStrings();
+
+        std::wstring envBlock;
+
+        if (envStrings == nullptr)
+        {
+            length = 2;
+            environmentVariables = new WCHAR[length];
+            environmentVariables[0] = L'\0';
+            environmentVariables[1] = L'\0';
+            return;
+        }
+
+        for (LPWCH env = envStrings; *env != L'\0'; env += wcslen(env) + 1)
+        {
+            if (wcsncmp(env, L"DD_", 3) == 0)
+            {
+                envBlock.append(env);
+                envBlock.push_back('\0');
+            }
+        }
+
+        // Environment block ends with a double null terminator
+        envBlock.push_back('\0');
+
+        if (envBlock.length() == 1)
+        {
+            // If the environment block was empty, we're still missing one null terminator
+            envBlock.push_back('\0');
+        }
+
+        length = static_cast<int32_t>(envBlock.length());
+        environmentVariables = new WCHAR[length];
+        memcpy(environmentVariables, envBlock.c_str(), length * sizeof(WCHAR));
+    }
+
+    // Concatenate two environment blocks
+    LPWSTR ConcatenateEnvironmentBlocks(LPCWSTR envBlock1, LPCWSTR envBlock2) {
+        // First, calculate the total length needed
+        size_t lenFirstBlock = 0;
+        size_t lenSecondBlock = 0;
+        LPCWSTR p;
+
+        // Calculate length of the first environment block
+        p = envBlock1;
+        while (*p) {
+            size_t len = wcslen(p) + 1; // Length of the entry including null terminator
+            lenFirstBlock += len;
+            p += len;
+        }
+
+        // Calculate length of the second environment block
+        p = envBlock2;
+        while (*p) {
+            size_t len = wcslen(p) + 1;
+            lenSecondBlock += len;
+            p += len;
+        }
+
+        // Allocate buffer for the merged environment block, add one for the final null character (double null termination)
+        LPWSTR mergedEnvBlock = (LPWSTR)malloc((lenFirstBlock + lenSecondBlock + 1) * sizeof(WCHAR));
```

Comment:
This is not a "merge" of the 2 env vars: are we sure that Windows will be ok with duplicated env vars?
And the first one is the dead process one which will have "priority" over the wer.exe ones.
Also, use new instead of malloc to have the same allocator in the code

Reply:
The first one having priority is done on purpose. I rewrote the method to be an actual merge